How many late days are you using for this assignment? 

In [74]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [75]:
# Load the given datasets
categorical_data = pd.read_csv('Chronic Kidney Disease Categorical.csv')
numerical_data = pd.read_csv('Chronic Kidney Disease Numerical.csv')

# Print the data
print("Chronic Kidney Disease Categorical Data:")
print(categorical_data)

print("\nChronic Kidney Disease Numerical Data:")
print(numerical_data)



Chronic Kidney Disease Categorical Data:
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
5       114717  0.0  0.0    normal    normal  notpresent  notpresent   no   
6       950823  4.0  0.0       NaN    normal  notpresent  notpresent   no   
7       901528  0.0  0.0    normal    normal  notpresent  notpresent   no   
8       433613  0.0  0.0    normal    normal  notpresent  notpresent   no   
9       621332  1.0  0.0  abnormal    normal  notpresent  notpresent   no   
10      708255  2.0  0.0    normal    normal  notpresent  notpresent   no   
11      365669  1.0  0.0    normal 

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The dataset contains medical records related to chronic kidney disease (CKD) with a target label indicating whether a patient has CKD ("ckd") or not ("notckd"). The features include both categorical and numerical variables: categorical features represent health indicators such as the presence of red blood cells (`rbc`), protein in urine (`pc`), and hypertension (`htn`), while numerical features include measurements like `al` (potentially albumin levels) and `su` (possibly specific gravity). For example, `rbc` and `pc` can have values such as "normal" or "abnormal," while `al` and `su` are represented by numerical values reflecting medical assessments. In total, categorical columns consist of `rbc`, `pc`, and `htn`, whereas numerical columns include `al`, `su`, and others that may apply. This blend of data is utilized to predict kidney health outcomes based on various clinical indicators.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [76]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
# Step 1: Get the initial shape of the numerical dataset and print it
initial_shape_numerical = numerical_data.shape
print(f"Original shape of the numerical dataset: {initial_shape_numerical}")

# Step 2: Find all unique IDs and count the duplicates
total_unique_ids_numerical = numerical_data['unique_id'].nunique()
print(f"\nTotal unique IDs in the numerical dataset: {total_unique_ids_numerical}")

# Count occurrences of each unique ID to find duplicates
duplicate_counts_numerical = numerical_data['unique_id'].value_counts()
duplicates_with_counts_numerical = duplicate_counts_numerical[duplicate_counts_numerical > 1]

# Print details about IDs with duplicates
print(f"\nUnique IDs with duplicates in the numerical dataset (count):\n{duplicates_with_counts_numerical}")

# Check how many actual duplicate rows exist
total_duplicate_rows = duplicate_counts_numerical.sum() - total_unique_ids_numerical
print(f"\nTotal number of actual duplicates in the numerical dataset: {total_duplicate_rows}")

# Step 3: Remove all duplicate rows, keeping only one of each unique ID
numerical_data_cleaned = numerical_data.drop_duplicates(subset='unique_id', keep='first')

# Validate removal of duplicates
# Check the number of unique IDs in the cleaned dataset
unique_ids_after_cleaning = numerical_data_cleaned['unique_id'].nunique()
print(f"Total unique IDs in the cleaned numerical dataset: {unique_ids_after_cleaning}")

# Step 4: Get the shape of the cleaned dataset and print it
new_shape_numerical = numerical_data_cleaned.shape
print(f"\nShape of the numerical dataset after removing duplicates: {new_shape_numerical}")

# Print the original shape again for comparison
print(f"\nOriginal shape of the numerical dataset: {initial_shape_numerical}")



# Repeat the same for categorical dataset. Print the duplicate rows and drop them
# Step 1: Get the initial shape of the categorical dataset and print it
initial_shape_categorical = categorical_data.shape
print(f"Original shape of the categorical dataset: {initial_shape_categorical}")

# Step 2: Find all unique IDs and count the duplicates
total_unique_ids_categorical = categorical_data['unique_id'].nunique()
print(f"\nTotal unique IDs in the categorical dataset: {total_unique_ids_categorical}")

# Count occurrences of each unique ID to find duplicates
duplicate_counts_cat = categorical_data['unique_id'].value_counts()
duplicates_with_counts_cat = duplicate_counts_cat[duplicate_counts_cat > 1]

# Print details about IDs with duplicates
print(f"\nUnique IDs with duplicates in the categorical dataset (count):\n{duplicates_with_counts_cat}")

# Check how many actual duplicate rows exist
total_duplicate_rows_cat = duplicate_counts_cat.sum() - total_unique_ids_categorical
print(f"\nTotal number of actual duplicates in the categorical dataset: {total_duplicate_rows_cat}")

# Step 3: Remove all duplicate rows, keeping only one of each unique ID
categorical_data_cleaned = categorical_data.drop_duplicates(subset='unique_id', keep='first')

# Validate removal of duplicates
# Check the number of unique IDs in the cleaned dataset
unique_ids_after_cleaning_cat = categorical_data_cleaned['unique_id'].nunique()
print(f"Total unique IDs in the cleaned categorical dataset: {unique_ids_after_cleaning_cat}")

# Step 4: Get the shape of the cleaned dataset and print it
new_shape_categorical = categorical_data_cleaned.shape
print(f"\nShape of the categorical dataset after removing duplicates: {new_shape_categorical}")

# Print the original shape again for comparison
print(f"\nOriginal shape of the categorical dataset: {initial_shape_categorical}")



Original shape of the numerical dataset: (420, 12)

Total unique IDs in the numerical dataset: 390

Unique IDs with duplicates in the numerical dataset (count):
unique_id
856244    3
118191    2
153227    2
862719    2
158929    2
572140    2
628271    2
415906    2
583971    2
354223    2
202793    2
337834    2
769512    2
843449    2
687021    2
643007    2
840392    2
441170    2
431871    2
708255    2
302906    2
283427    2
501321    2
872034    2
721270    2
261137    2
224481    2
240975    2
938027    2
Name: count, dtype: int64

Total number of actual duplicates in the numerical dataset: 30
Total unique IDs in the cleaned numerical dataset: 390

Shape of the numerical dataset after removing duplicates: (390, 12)

Original shape of the numerical dataset: (420, 12)
Original shape of the categorical dataset: (430, 14)

Total unique IDs in the categorical dataset: 400

Unique IDs with duplicates in the categorical dataset (count):
unique_id
856244    3
708255    2
224481    2
58

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

In this analysis, I chose to perform an inner join to merge the numerical and categorical datasets based on their `unique_id` values. This method ensures that only records with matching IDs in both datasets are included in the combined dataset, thereby maintaining data integrity and completeness essential for accurate analysis and effective machine learning model training.

In [77]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
# Step 1: Get the initial shape of the numerical dataset
initial_shape_numerical = numerical_data_cleaned.shape

# Step 2: Get the initial shape of the categorical dataset
initial_shape_categorical = categorical_data_cleaned.shape

# Step 3: Count unique IDs in each dataset
unique_ids_numerical = numerical_data_cleaned['unique_id'].nunique()
unique_ids_categorical = categorical_data_cleaned['unique_id'].nunique()

# Step 4: Merge the datasets on unique_id
combined_dataset = pd.merge(numerical_data_cleaned, categorical_data_cleaned, on='unique_id', how='inner')

# Step 5: Count unique IDs in the combined dataset
unique_ids_combined = combined_dataset['unique_id'].nunique()

# Step 6: Print summary information
print(f"Unique IDs in the numerical dataset: {unique_ids_numerical}")
print(f"Unique IDs in the categorical dataset: {unique_ids_categorical}")
print(f"Unique IDs in the combined dataset: {unique_ids_combined}")
print(f"Shape of the numerical dataset: {initial_shape_numerical}")
print(f"Shape of the categorical dataset: {initial_shape_categorical}")
print(f"Shape of the combined dataset: {combined_dataset.shape}")

#Print the combined dataset
print("\nCombined Dataset:")
print(combined_dataset)


Unique IDs in the numerical dataset: 390
Unique IDs in the categorical dataset: 400
Unique IDs in the combined dataset: 390
Shape of the numerical dataset: (390, 12)
Shape of the categorical dataset: (400, 14)
Shape of the combined dataset: (390, 25)

Combined Dataset:
     unique_id   age     bp    bgr     bu     sc    sod   pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0   0.50  147.0   3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0   1.00  144.0   5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0   0.80  135.0   4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0   4.00  136.0   4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0   0.90  140.0   4.1   NaN   NaN   
5       992643  35.0   70.0   82.0   36.0   1.10  150.0   3.5  14.5  52.0   
6       573400  60.0   80.0    NaN    NaN    NaN    NaN   NaN   NaN   NaN   
7       109053  64.0   60.0  106.0   27.0   0.70  150.0   3.3  14.4  42.0   
8       352995  33.0   60.0  130.0   

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [78]:
# Step 1: Calculate the percentage of rows containing at least one missing value
percentage_missing = combined_dataset.isnull().any(axis=1).mean() * 100

# Step 2: Print the percentage of rows with at least one missing value
print(f"Percentage of rows with at least one missing value: {percentage_missing:.2f}%")

# Step 3: Drop rows with missing values using boolean indexing
cleaned_combined_dataset = combined_dataset[~combined_dataset.isnull().any(axis=1)]

# Step 4: Print the shape of both the original and cleaned datasets
print(f"Shape of the original combined dataset: {combined_dataset.shape}")
print(f"Shape of the cleaned combined dataset: {cleaned_combined_dataset.shape}")

# Step 5: Print the cleaned dataset
print("\nCleaned Combined Dataset (after dropping rows with missing values):")
print(cleaned_combined_dataset)


Percentage of rows with at least one missing value: 60.77%
Shape of the original combined dataset: (390, 25)
Shape of the cleaned combined dataset: (153, 25)

Cleaned Combined Dataset (after dropping rows with missing values):
     unique_id   age     bp    bgr     bu    sc    sod   pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0   19.0   0.5  147.0   3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0   25.0   1.0  144.0   5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0   35.0   0.8  135.0   4.1  16.2  50.0  ...   
7       109053  64.0   60.0  106.0   27.0   0.7  150.0   3.3  14.4  42.0  ...   
13      349892  59.0   70.0  424.0   55.0   1.7  138.0   4.5  12.6  37.0  ...   
17      630177  58.0   80.0  131.0   18.0   1.1  141.0   3.5  15.8  53.0  ...   
18      741276  23.0   60.0   95.0   24.0   0.8  145.0   5.0  15.0  52.0  ...   
19      546010  49.0   80.0  122.0   32.0   1.2  139.0   3.9  17.0  41.0  ...   
20      460465  47.0   60.0  109.0   25.0   

## Part 3.4: Sort the dataset according to the Labels

In [79]:
# Step 1: Sort the cleaned dataset by the 'Target' column
sorted_dataset = cleaned_combined_dataset.sort_values(by='Target')

# Step 2: Reset the index of the sorted dataset
sorted_dataset.reset_index(drop=True, inplace=True)

# Step 3: Print the sorted dataset
print("\nSorted Dataset by 'Target':")
print(sorted_dataset)



Sorted Dataset by 'Target':
     unique_id   age     bp    bgr     bu    sc    sod   pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0   1.7  125.0   3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0   5.6  140.0   2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0   7.2  114.0   3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0   3.9  135.0   5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0   2.3  136.0   5.2  11.1  33.0  ...   
5       566924  49.0  100.0  129.0  158.0  11.8  122.0   3.2   8.1  24.0  ...   
6       546225  46.0   60.0  163.0   92.0   3.3  141.0   4.0   9.8  28.0  ...   
7       602404  55.0   90.0  273.0  235.0  14.2  132.0   3.4   8.3  22.0  ...   
8       514721  56.0   90.0  129.0  107.0   6.7  131.0   4.8   9.1  29.0  ...   
9       899916  83.0   70.0  102.0   60.0   2.6  115.0   5.7   8.7  26.0  ...   
10      621332  59.0   80.0  303.0   35.0   1.3  122.0   3.5  10.4  35.0  ...   

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [80]:
# Write code here
# Define the categorical columns to be transformed
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

# Step 1: Create dummy variables explicitly for 'rbc' and other binary columns
for col in categorical_columns:
    if col in cleaned_combined_dataset.columns and cleaned_combined_dataset[col].dtype == 'object':
        # Create dummy variables with drop_first=False to keep both categories
        dummies = pd.get_dummies(cleaned_combined_dataset[col], prefix=col, drop_first=False)
        
        # Convert boolean values to integers
        dummies = dummies.astype(int)
        
        # Combine the new dummy columns with the original dataset
        cleaned_combined_dataset = pd.concat([cleaned_combined_dataset, dummies], axis=1)

# Step 2: Remove the original categorical columns if they are still in the DataFrame
# Check if the columns exist before dropping
existing_columns_to_drop = [col for col in categorical_columns if col in cleaned_combined_dataset.columns]
cleaned_combined_dataset.drop(columns=existing_columns_to_drop, inplace=True)

# Step 3: Print the transformed dataset
print("\nDataset after converting categorical columns to dummy variables:")
print(cleaned_combined_dataset)


Dataset after converting categorical columns to dummy variables:
     unique_id   age     bp    bgr     bu    sc    sod   pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0   19.0   0.5  147.0   3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0   25.0   1.0  144.0   5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0   35.0   0.8  135.0   4.1  16.2  50.0  ...   
7       109053  64.0   60.0  106.0   27.0   0.7  150.0   3.3  14.4  42.0  ...   
13      349892  59.0   70.0  424.0   55.0   1.7  138.0   4.5  12.6  37.0  ...   
17      630177  58.0   80.0  131.0   18.0   1.1  141.0   3.5  15.8  53.0  ...   
18      741276  23.0   60.0   95.0   24.0   0.8  145.0   5.0  15.0  52.0  ...   
19      546010  49.0   80.0  122.0   32.0   1.2  139.0   3.9  17.0  41.0  ...   
20      460465  47.0   60.0  109.0   25.0   1.1  141.0   4.7  15.8  41.0  ...   
22      965166  48.0   80.0   75.0   22.0   0.8  137.0   5.0  16.8  51.0  ...   
23      201677  33.0   80.0   89.0   19.0  

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: When dealing with binary variables, using a single column to indicate the presence (1) or absence (0) of a category is equivalent to creating two separate dummy variables. This approach simplifies the model without losing information. However, for categorical features with more than two values, such as "type of condiment" that can take five values, it is crucial to use dummy variables (one-hot encoding) instead of a single numeric column. This ensures that the model accurately interprets the categorical variable without imposing an artificial order. Using dummy variables allows the model to treat each category independently, enhancing its ability to learn and make predictions.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [81]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# Step 1: Create a copy of the dataset to work with
cleaned_dataset_no_outliers = cleaned_combined_dataset.copy()

# Step 2: Loop through each numerical column to identify and drop outliers
for column in numerical_columns:
    if column in cleaned_dataset_no_outliers.columns:
        mean = cleaned_dataset_no_outliers[column].mean()
        std_dev = cleaned_dataset_no_outliers[column].std()
        
        # Define lower and upper bounds for outliers
        lower_bound = mean - 3 * std_dev
        upper_bound = mean + 3 * std_dev
        
        # Filter out the outliers
        cleaned_dataset_no_outliers = cleaned_dataset_no_outliers[(cleaned_dataset_no_outliers[column] >= lower_bound) & 
                                                                  (cleaned_dataset_no_outliers[column] <= upper_bound)]

# Step 3: Print the cleaned dataset without outliers
print("\nCleaned Dataset (after removing outliers):")
print(cleaned_dataset_no_outliers)



Cleaned Dataset (after removing outliers):
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
7       109053  64.0   60.0  106.0   27.0  0.7  150.0  3.3  14.4  42.0  ...   
17      630177  58.0   80.0  131.0   18.0  1.1  141.0  3.5  15.8  53.0  ...   
18      741276  23.0   60.0   95.0   24.0  0.8  145.0  5.0  15.0  52.0  ...   
19      546010  49.0   80.0  122.0   32.0  1.2  139.0  3.9  17.0  41.0  ...   
20      460465  47.0   60.0  109.0   25.0  1.1  141.0  4.7  15.8  41.0  ...   
22      965166  48.0   80.0   75.0   22.0  0.8  137.0  5.0  16.8  51.0  ...   
23      201677  33.0   80.0   89.0   19.0  1.1  144.0  5.0  15.0  40.0  ...   
24      976875  33.0   60.0   80.0   25.0  0.9  146.0  3.5  14.1  48.0  ...   
35      

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [82]:
from sklearn.preprocessing import StandardScaler

# Define the numerical columns
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# Step 1: Initialize the StandardScaler
scaler = StandardScaler()

# Step 2: Fit the scaler to the numerical columns and transform the data using the cleaned dataset without outliers
cleaned_dataset_no_outliers[numerical_columns] = scaler.fit_transform(cleaned_dataset_no_outliers[numerical_columns])

# Step 3: Print the standardized dataset
print("\nStandardized Dataset (after normalization):")
print(cleaned_dataset_no_outliers)



Standardized Dataset (after normalization):
     unique_id       age        bp       bgr        bu        sc       sod  \
0       203694 -0.640847  0.730553 -0.495477 -0.826391 -0.602214  1.152966   
1       938027 -0.325404 -1.314996 -0.295785 -0.601243 -0.265116  0.624799   
2       421471 -0.703936 -1.314996 -0.229221 -0.225995 -0.399955 -0.959702   
7       109053  0.999456 -1.314996 -0.340161 -0.526193 -0.467375  1.681133   
17      630177  0.620924  0.730553  0.214540 -0.863916 -0.197696  0.096632   
18      741276 -1.587175 -1.314996 -0.584229 -0.638768 -0.399955  0.800855   
19      546010  0.053127  0.730553  0.014848 -0.338569 -0.130277 -0.255479   
20      460465 -0.073050 -1.314996 -0.273597 -0.601243 -0.197696  0.096632   
22      965166 -0.009961  0.730553 -1.027990 -0.713817 -0.399955 -0.607591   
23      201677 -0.956290  0.730553 -0.717358 -0.826391 -0.197696  0.624799   
24      976875 -0.956290 -1.314996 -0.917050 -0.601243 -0.332535  0.976910   
35      431871  0.5

## Part 3.8: Remove Unnecessary columns

Are there any columns in  this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

Below checks for constant columns to be excluded

In [83]:
# Initialize a list to store constant columns
constant_columns = []

# Check each column in the cleaned dataset without outliers
for column in cleaned_dataset_no_outliers.columns:
    # Check if the number of unique values in the column is 1
    if cleaned_dataset_no_outliers[column].nunique() <= 1:
        constant_columns.append(column)

# Print the constant columns, if any
if constant_columns:
    print("Constant Columns (where all values are the same):")
    print(constant_columns)
else:
    print("No constant columns found in the dataset.")


No constant columns found in the dataset.


Because there are no constant columns, we will just remove the unique ID column because it has no predictive value for a machine learning application.

In [84]:
# Check if 'unique_id' is in the dataset without outliers and remove it if it is
if 'unique_id' in cleaned_dataset_no_outliers.columns:
    cleaned_dataset_no_outliers.drop(columns=['unique_id'], inplace=True)
    print("The 'unique_id' column has been removed from the dataset.")
else:
    print("The 'unique_id' column was not found in the dataset.")

# Print the updated dataset to verify the removal
print("\nUpdated Dataset (after removing 'unique_id'):")
print(cleaned_dataset_no_outliers)


The 'unique_id' column has been removed from the dataset.

Updated Dataset (after removing 'unique_id'):
          age        bp       bgr        bu        sc       sod       pot  \
0   -0.640847  0.730553 -0.495477 -0.826391 -0.602214  1.152966 -1.303189   
1   -0.325404 -1.314996 -0.295785 -0.601243 -0.265116  0.624799  1.058176   
2   -0.703936 -1.314996 -0.229221 -0.225995 -0.399955 -0.959702 -0.358643   
7    0.999456 -1.314996 -0.340161 -0.526193 -0.467375  1.681133 -1.618038   
17   0.620924  0.730553  0.214540 -0.863916 -0.197696  0.096632 -1.303189   
18  -1.587175 -1.314996 -0.584229 -0.638768 -0.399955  0.800855  1.058176   
19   0.053127  0.730553  0.014848 -0.338569 -0.130277 -0.255479 -0.673492   
20  -0.073050 -1.314996 -0.273597 -0.601243 -0.197696  0.096632  0.585903   
22  -0.009961  0.730553 -1.027990 -0.713817 -0.399955 -0.607591  1.058176   
23  -0.956290  0.730553 -0.717358 -0.826391 -0.197696  0.624799  1.058176   
24  -0.956290 -1.314996 -0.917050 -0.601243 -0.3

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [85]:
# Export the cleaned DataFrame (with outliers removed) to a new CSV file
cleaned_dataset_no_outliers.to_csv('final_cleaned_dataset.csv', index=False)
print("DataFrame has been exported to 'final_cleaned_dataset.csv'.")

# Export the cleaned DataFrame (with outliers removed) to a new JSON file
cleaned_dataset_no_outliers.to_json('final_cleaned_dataset.json', orient='records', lines=True)
print("DataFrame has been exported to 'final_cleaned_dataset.json'.")


DataFrame has been exported to 'final_cleaned_dataset.csv'.
DataFrame has been exported to 'final_cleaned_dataset.json'.


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [86]:
# Print the first 15 rows of the cleaned dataset without outliers
print("First 15 rows of the cleaned dataset (after outlier removal):")
print(cleaned_dataset_no_outliers.head(15))


First 15 rows of the cleaned dataset (after outlier removal):
         age        bp       bgr        bu        sc       sod       pot  \
0  -0.640847  0.730553 -0.495477 -0.826391 -0.602214  1.152966 -1.303189   
1  -0.325404 -1.314996 -0.295785 -0.601243 -0.265116  0.624799  1.058176   
2  -0.703936 -1.314996 -0.229221 -0.225995 -0.399955 -0.959702 -0.358643   
7   0.999456 -1.314996 -0.340161 -0.526193 -0.467375  1.681133 -1.618038   
17  0.620924  0.730553  0.214540 -0.863916 -0.197696  0.096632 -1.303189   
18 -1.587175 -1.314996 -0.584229 -0.638768 -0.399955  0.800855  1.058176   
19  0.053127  0.730553  0.014848 -0.338569 -0.130277 -0.255479 -0.673492   
20 -0.073050 -1.314996 -0.273597 -0.601243 -0.197696  0.096632  0.585903   
22 -0.009961  0.730553 -1.027990 -0.713817 -0.399955 -0.607591  1.058176   
23 -0.956290  0.730553 -0.717358 -0.826391 -0.197696  0.624799  1.058176   
24 -0.956290 -1.314996 -0.917050 -0.601243 -0.332535  0.976910 -1.303189   
35  0.557836 -1.314996  0.

Paste here:
|    age     |     bp      |    bgr      |     bu      |    sc      |     sod     |     pot     |    hemo     |     pcv     |     wbcc    | ... | cad_no | cad_yes | appet_good |
|------------|-------------|-------------|-------------|------------|-------------|-------------|--------------|-------------|-------------|-----|--------|---------|------------|
| -0.640847  |  0.730553   | -0.495477   | -0.826391   | -0.602214  |  1.152966   | -1.303189   | -0.346119    | -0.011966   | -0.322121   | ... |   1    |    0    |     1      |
| -0.325404  | -1.314996   | -0.295785   | -0.601243   | -0.265116  |  0.624799   |  1.058176   |  1.481310    | -0.156643   | -0.366426   | ... |   1    |    0    |     1      |
| -0.703936  | -1.314996   | -0.229221   | -0.225995   | -0.399955  | -0.959702   | -0.358643   |  0.785147    |  0.856096   | -1.119597   | ... |   1    |    0    |     1      |
|  0.999456  | -1.314996   | -0.340161   | -0.526193   | -0.467375  |  1.681133   | -1.618038   |  0.001963    | -0.301320   |  0.032312   | ... |   1    |    0    |     1      |
|  0.620924  |  0.730553   |  0.214540   | -0.863916   | -0.197696  |  0.096632   | -1.303189   |  0.611106    |  1.290126   | -0.543642   | ... |   1    |    0    |     1      |
| -1.587175  | -1.314996   | -0.584229   | -0.638768   | -0.399955  |  0.800855   |  1.058176   |  0.263024    |  1.145449   | -0.765163   | ... |   1    |    0    |     1      |
|  0.053127  |  0.730553   |  0.014848   | -0.338569   | -0.130277  | -0.255479   | -0.673492   |  0.100817    |  0.644820   | -0.004461   | ... |   1    |    0    |     1      |
| -0.073050  | -1.314996   | -0.273597   | -0.601243   | -0.197696  |  0.096632   |  0.585903   |  0.116824    |  0.728104   | -0.129353   | ... |   1    |    0    |     1      |
| -0.009961  |  0.730553   | -1.027990   | -0.713817   | -0.399955  | -0.607591   |  1.058176   |  0.003728    | -0.019763   | -0.254904   | ... |   1    |    0    |     1      |
| -0.956290  |  0.730553   | -0.717358   | -0.826391   | -0.197696  |  0.624799   |  1.058176   |  0.335215    |  0.852895   | -0.823819   | ... |   1    |    0    |     1      |
| -0.956290  | -1.314996   | -0.917050   | -0.601243   | -0.332535  |  0.976910   | -1.303189   |  0.278456    | -0.046162   | -0.362904   | ... |   1    |    0    |     1      |
|  0.557836  | -1.314996   |  0.236728   | -0.863916   | -0.197696  |  1.681133   |  0.585903   |  0.883456    |  0.116824   | -0.731019   | ... |   1    |    0    |     1      |
| -0.388493  |  0.730553   |  0.236728   | -0.638768   | -0.467375  | -0.079424   | -0.358643   |  0.720001    |  0.122030   | -0.269522   | ... |   1    |    0    |     1      |
|  1.567253  | -1.314996   | -0.061306   | -0.096718   | -0.554256  |  1.495111   | -0.325113   |  0.341315    |  0.536728   | -0.150301   | ... |   1    |    0    |     1      |


### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer: To mitigate data conversion hallucinations from language model (LLM) conversions, particularly in larger datasets, it’s essential to implement several strategies. First, clear specifications for data formats should be maintained to ensure consistency, along with validation checks and manual reviews of a sample of converted data to catch discrepancies. Automated tests can be established to verify transformations and assert expected properties post-conversion, while chunk processing allows for easier tracking of errors. Involving subject matter experts for complex transformations enhances the accuracy of conversions, and using established libraries or tools provides additional verification of data integrity. By combining these approaches, the risk of inaccuracies during data conversions can be significantly reduced.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [87]:

# Step 1: Read the cleaned dataset from the CSV file
final_cleaned_dataset = pd.read_csv('final_cleaned_dataset.csv')

# Step 2: Define the mapping of old column names (abbreviations) to new column names (full names)
column_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Step 3: Function to replace abbreviations in column names
def replace_abbreviations(column_name):
    # Replace standalone abbreviations and abbreviations in compound names
    for abbr, full_name in column_mapping.items():
        # Replace standalone abbreviation
        if f"_{abbr}" in column_name:  # Handle compound names with underscore
            column_name = column_name.replace(f"_{abbr}", f"_{full_name}")
        if column_name.startswith(abbr + "_"):  # Handle case like abbr_name
            column_name = column_name.replace(abbr + "_", full_name + "_")
        if column_name == abbr:  # Handle standalone abbreviation
            column_name = full_name
    return column_name

# Step 4: Apply the function to rename the columns
final_cleaned_dataset.columns = [replace_abbreviations(col) for col in final_cleaned_dataset.columns]
# Step 1: Print the shape of the DataFrame to see the total number of rows and columns
print(f"Shape of the final cleaned DataFrame: {final_cleaned_dataset.shape}")

# Step 2: Print the total number of rows
print(f"Total number of rows in the final cleaned DataFrame: {len(final_cleaned_dataset)}")

# Step 3: Optionally, print all rows (or a specified number of rows) to visually inspect the data
# Adjust the display settings to ensure full visibility of rows
pd.set_option('display.max_rows', None)  # Set None to display all rows

# Print the entire DataFrame (use with caution if DataFrame is large)
print("\nComplete Dataset:")
print(final_cleaned_dataset)


Shape of the final cleaned DataFrame: (133, 35)
Total number of rows in the final cleaned DataFrame: 133

Complete Dataset:
          age  blood pressure  blood glucose random  blood urea  \
0   -0.640847        0.730553             -0.495477   -0.826391   
1   -0.325404       -1.314996             -0.295785   -0.601243   
2   -0.703936       -1.314996             -0.229221   -0.225995   
3    0.999456       -1.314996             -0.340161   -0.526193   
4    0.620924        0.730553              0.214540   -0.863916   
5   -1.587175       -1.314996             -0.584229   -0.638768   
6    0.053127        0.730553              0.014848   -0.338569   
7   -0.073050       -1.314996             -0.273597   -0.601243   
8   -0.009961        0.730553             -1.027990   -0.713817   
9   -0.956290        0.730553             -0.717358   -0.826391   
10  -0.956290       -1.314996             -0.917050   -0.601243   
11   0.557836       -1.314996              0.236728   -0.863916   
12  -

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [88]:
# Step 1: Read the cleaned dataset from the CSV file (if not already done)
# final_cleaned_dataset = pd.read_csv('final_cleaned_dataset.csv')  # Uncomment if reading from CSV

# Step 2: Print the current columns to confirm the target column names
print("Columns in the final cleaned dataset:")
print(final_cleaned_dataset.columns)

# Step 3: Group the DataFrame by both target columns and count occurrences
target_columns = ['Target_ckd', 'Target_notckd']
if all(col in final_cleaned_dataset.columns for col in target_columns):
    # Create a new DataFrame to count occurrences
    result = final_cleaned_dataset[target_columns].copy()
    result['count'] = 1  # Create a count column for grouping
    aggregated_result = result.groupby(target_columns).count().reset_index()
    
    # Step 4: Print the result
    print("\nCounts for each target category:")
    print(aggregated_result)
else:
    print(f"One or more of the target columns {target_columns} do not exist in the dataset.")


Columns in the final cleaned dataset:
Index(['age', 'blood pressure', 'blood glucose random', 'blood urea',
       'serum creatinine', 'sodium', 'potassium', 'hemoglobin',
       'packed cell volume', 'wbcc', 'rbcc', 'albumin', 'sugar',
       'red blood cells_abnormal', 'red blood cells_normal',
       'pus cell_abnormal', 'pus cell_normal', 'pus cell clumps_notpresent',
       'pus cell clumps_present', 'bacteria_notpresent', 'bacteria_present',
       'hypertension_no', 'hypertension_yes', 'diabetes mellitus_no',
       'diabetes mellitus_yes', 'coronary artery disease_no',
       'coronary artery disease_yes', 'appetite_good', 'appetite_poor',
       'pedal edema_no', 'pedal edema_yes', 'anemia_no', 'anemia_yes',
       'Target_ckd', 'Target_notckd'],
      dtype='object')

Counts for each target category:
   Target_ckd  Target_notckd  count
0           0              1    111
1           1              0     22


*Describe what the above code does here*


The provided code begins by checking the column names in the `final_cleaned_dataset` DataFrame to ensure that the target columns—`Target_ckd` and `Target_notckd`—exist. It then creates a new DataFrame that includes these target columns and adds a count column initialized to 1 for use in grouping. The code subsequently groups the DataFrame according to the unique combinations of the target columns and counts the occurrences of each combination. Finally, it prints the aggregated counts, providing insights into the distribution of the target classes within the dataset. This process is essential for understanding how many instances correspond to each of the potential output classes prior to modeling.
